pkg/cloudprovider/controller/nodecontroller.go

```diff
@@ -446,54 +505,23 @@ func (nc *NodeController) MonitorNodeStatus() error {
 		var gracePeriod time.Duration
 		var lastReadyCondition api.NodeCondition
 		node := &nodes.Items[i]
-		readyCondition := nc.getCondition(node, api.NodeReady)
-		if readyCondition == nil {
-			// If ready condition is nil, then kubelet (or nodecontroller) never posted node status.
-			// A fake ready condition is created, where LastProbeTime and LastTransitionTime is set
-			// to node.CreationTimestamp to avoid handle the corner case.
-			lastReadyCondition = api.NodeCondition{
-				Type:               api.NodeReady,
-				Status:             api.ConditionUnknown,
-				LastProbeTime:      node.CreationTimestamp,
-				LastTransitionTime: node.CreationTimestamp,
-			}
-			gracePeriod = nodeStartupGracePeriod
-		} else {
-			// If ready condition is not nil, make a copy of it, since we may modify it in place later.
-			lastReadyCondition = *readyCondition
-			gracePeriod = nodeMonitorGracePeriod
-		}
-
-		// Check last time when NodeReady was updated.
-		if nc.now().After(lastReadyCondition.LastProbeTime.Add(gracePeriod)) {
-			// NodeReady condition was last set longer ago than gracePeriod, so update it to Unknown
-			// (regardless of its current value) in the master, without contacting kubelet.
-			if readyCondition == nil {
-				glog.V(2).Infof("node %v is never updated by kubelet")
-				node.Status.Conditions = append(node.Status.Conditions, api.NodeCondition{
-					Type:               api.NodeReady,
-					Status:             api.ConditionUnknown,
-					Reason:             fmt.Sprintf("Kubelet never posted node status"),
-					LastProbeTime:      node.CreationTimestamp,
-					LastTransitionTime: nc.now(),
-				})
-			} else {
-				glog.V(2).Infof("node %v hasn't been updated for %+v. Last ready condition is: %+v",
-					node.Name, nc.now().Time.Sub(lastReadyCondition.LastProbeTime.Time), lastReadyCondition)
-				if lastReadyCondition.Status != api.ConditionUnknown {
-					readyCondition.Status = api.ConditionUnknown
-					readyCondition.Reason = fmt.Sprintf("Kubelet stopped posting node status")
-					// LastProbeTime is the last time we heard from kubelet.
-					readyCondition.LastProbeTime = lastReadyCondition.LastProbeTime
-					readyCondition.LastTransitionTime = nc.now()
-				}
-			}
-			_, err = nc.kubeClient.Nodes().Update(node)
+		for rep := 0; rep < nodeStatusUpdateRetry; rep++ {
+			err, gracePeriod, lastReadyCondition = nc.tryUpdateNodeStatus(node)
 			if err != nil {
-				glog.Errorf("error updating node %s: %v", node.Name, err)
+				name := node.Name
+				node, err = nc.kubeClient.Nodes().Get(name)
+				if err != nil {
+					glog.Errorf("Failed while getting a Node to retry updating NodeStatus. Probably Node %s was deleted.", name)
+				}
+			} else {
+				break
 			}
 		}
+		if err != nil {
+			glog.Errorf("Update status  of Node %v from NodeController exceeds retry count. Skipping", node.Name)
```

Comment:
should you "continue" after this line? if the update failed, I'm not sure you want to execute the pod eviction logic. probably better to err on the safe side.